lib/msal-node/test/client/ManagedIdentitySources/Imds.spec.ts

```diff
@@ -170,6 +173,130 @@ describe("Acquires a token successfully via an IMDS Managed Identity", () => {
+
+                const sendGetRequestAsyncSpy: jest.SpyInstance = jest
+                    .spyOn(networkClient, <any>"sendGetRequestAsync")
+                    // override the networkClient's sendGetRequestAsync method to return a 500.
+                    // after this override, original functionality will be restored
+                    // and the network request will complete successfully
+                    .mockReturnValueOnce(
+                        managedIdentityNetworkErrorClient.sendGetRequestForRetryAsync()
+                    );
+
+                const networkManagedIdentityResult: AuthenticationResult =
+                    await managedIdentityApplication.acquireToken(
+                        managedIdentityRequestParams
+                    );
+
+                expect(sendGetRequestAsyncSpy).toHaveBeenCalledTimes(2);
+                expect(networkManagedIdentityResult.accessToken).toEqual(
+                    DEFAULT_USER_SYSTEM_ASSIGNED_MANAGED_IDENTITY_AUTHENTICATION_RESULT.accessToken
+                );
+
+                jest.restoreAllMocks();
+            });
+
+            test("returns a 500 error response from the network request permanently", async () => {
+                expect(ManagedIdentityTestUtils.isIMDS()).toBe(true);
+
+                const sendGetRequestAsyncSpy: jest.SpyInstance = jest
+                    .spyOn(networkClient, <any>"sendGetRequestAsync")
+                    // permanently override the networkClient's sendGetRequestAsync method to return a 500.
+                    .mockReturnValue(
+                        managedIdentityNetworkErrorClient.sendGetRequestForRetryAsync()
+                    );
+
+                let serverError: ServerError = new ServerError();
+                try {
+                    await managedIdentityApplication.acquireToken(
+                        managedIdentityRequestParams
+                    );
+                } catch (e) {
+                    serverError = e as ServerError;
+                }
+
+                expect(serverError.errorCode).toEqual(
+                    MANAGED_IDENTITY_TOKEN_RETRIEVAL_ERROR
+                );
+                expect(sendGetRequestAsyncSpy).toHaveBeenCalledTimes(2);
+
+                jest.restoreAllMocks();
+            });
+        });
+
+        describe("System Assigned", () => {
+            let managedIdentityApplication: ManagedIdentityApplication;
+            beforeEach(() => {
+                managedIdentityApplication = new ManagedIdentityApplication(
+                    systemAssignedConfig
+                );
+            });
+
+            test("returns a 500 error response from the network request, just the first time", async () => {
+                expect(ManagedIdentityTestUtils.isIMDS()).toBe(true);
+
+                const sendGetRequestAsyncSpy: jest.SpyInstance = jest
+                    .spyOn(networkClient, <any>"sendGetRequestAsync")
```

Comment:
Wouldn't it be better to use regular HTTP level mocking instead of mocking the API? With HTTP level mocking, you also test that your change does not affect the ability of the app developer to use their own `INetworkModule`

Reply:
I'm really confused by this comment. I AM mocking the http request. In fact, that's the only thing I'm mocking here: the return value from the http request. I do the same thing in the Azure Arc tests.

Reply:
Using the same Managed Identity Application, I need to mock one of the return values to simulate a 500 so that the request is retried.

Reply:
It's confusing because you are mocking `sendGetRequestAsync` - but which `sendGetRequestAsync` is it? 

- the httpClientWithNoRetries?
- the httpClientWithRetries?

Reply:
Ok, I looked at this and it's ok. I was thinking that the `managedIdentityNetworkErrorClient` can have a list (or a queue) of responses and your test can just check how many responses have been served. This is ok too.